Switch the MPS manual to using the Read the Docs theme

rptb1 opened the pull request:
Working on #157 by experimenting with using the Read the Docs theme.

See https://memory-pool-system--166.org.readthedocs.build/en/166/ for the manual built from this pull request by Read the Docs CI.

The search in the Read the Docs theme works when the manual is built locally.  It does not depend on the manual being published on a server.

This change could remove the MPS theme and thus reduce maintenance costs.

rptb1 commented:
Also consider adopting https://stackoverflow.com/a/66828887 on this branch.

rptb1 commented:
The current manual has a sidebar like this:
![image](https://user-images.githubusercontent.com/419807/220527970-db024297-774b-467e-a860-30d3c031e317.png)
sourced from [manual/source/_templates](https://github.com/Ravenbrook/mps/tree/17c57ad0b7d9b95640c2fd53450b719d9b79f384/manual/source/_templates) .  It does not appear in the Read the Docs theme.  We must consider whether it is worth retaining and if so, how to do it.  (It indicentally contains URLs to www.ravenbrook.com that are going out of date due to Git migration.)

rptb1 commented:
We need to examine the manual for cases where the MPS Sphinx extension and the MPS theme interact.  For example, in the [Bibiolgraphy](https://memory-pool-system--166.org.readthedocs.build/en/166/bib.html), abstracts of papers now render with a strongly-highlighted exclamation-mark box like this
![image](https://user-images.githubusercontent.com/419807/220597811-b78f32bd-e500-44f3-86d6-bbae1c051ee9.png)
as opposed to a more gentle call-out like this in the MPS theme:
![image](https://user-images.githubusercontent.com/419807/220598036-4510d4da-a76c-4bf7-a763-f53ed821633e.png)

Of course we should try to use whatever intentional markup Sphinx provides for any purpose.

rptb1 commented:
> Of course we should try to use whatever intentional markup Sphinx provides for any purpose.

In the case of abstracts, an admonition is used: https://github.com/Ravenbrook/mps/blob/17c57ad0b7d9b95640c2fd53450b719d9b79f384/manual/source/bib.rst?plain=1#L10

In solving this problem, we should:
- look at conventions and extensions used in Sphinx for bibliographies
- ensure the bibliography comes out well in e.g. LaTeX output from Sphinx
- not break the [bibliography at Memory Management Reference](https://www.memorymanagement.org/bib.html#bibliography) or any links to it from the web

rptb1 commented:
Switching themes may fix #121 but this needs checking.

rptb1 commented:
> Switching themes may fix #121 but this needs checking.

Checked.  The layout problem does not appear with Sphinx 6.1.3 with the Read the Docs theme, the [latest available from PyPI](https://pypi.org/project/Sphinx/).

rptb1 commented:
The content navigation bar on the left in the Read the Docs theme is incorrect.

1. In the Read the Docs theme, the entry "Home" and subsequent entries seem to be renderings of the Memory Management Reference.
2. The index does not appear.

Here are the entries using the Read the Docs theme:

    Guide
    Reference
    Pool reference
    Design
    Old design

    Bibliography
    Memory Management Glossary
    Index to source code
    Memory Pool System Kit Open Source License
    Contact us
    Contributing to the MPS
    Release notes

    Introduction to memory management
    Home
    Frequently Asked Questions
    Copyright
    Acknowledgements

Compare to the way the contents appear in the MPS Sphinx theme:

    Guide
    Reference
    Pool reference
    Design
    Old design

    Bibliography
    Memory Management Glossary
    Index to source code
    Memory Pool System Kit Open Source License
    Contact us
    Contributing to the MPS
    Release notes

    Index

rptb1 commented:
> The content navigation bar on the left in the Read the Docs theme is incorrect.

This is no doubt because the Read the Docs theme does not support this configuration https://github.com/Ravenbrook/mps/blob/96a1cb634364e59142e3a3b879d82db6652d0f67/manual/source/conf.py#L41-L43

See also [How the table of contents displays](https://sphinx-rtd-theme.readthedocs.io/en/stable/configuring.html#how-the-table-of-contents-displays).

rptb1 commented:
> The content navigation bar on the left in the Read the Docs theme is incorrect.

It looks like the Read the Docs theme is ignoring the index inserted here, and also the `:hidden:` instruction.
https://github.com/Ravenbrook/mps/blob/96a1cb634364e59142e3a3b879d82db6652d0f67/manual/source/index.rst#L28-L38

rptb1 commented:
> It looks like the Read the Docs theme is ignoring the index inserted here

This can be solved by extending a Jinja template layout.html, as described here https://stackoverflow.com/a/37843854 .

rptb1 commented:
This branch is in danger of breaking the build of the Memory Management Reference described here https://github.com/Ravenbrook/mps/issues/165#issuecomment-1447342878 .

rptb1 commented:
> It looks like the Read the Docs theme is ignoring [...] the `:hidden:` instruction.

Fixed in https://github.com/Ravenbrook/mps/pull/166/commits/a708cdba26f3c4db0404a33e12bfa6d3069a0c67

rptb1 commented:
> This can be solved by extending a Jinja template layout.html, as described here https://stackoverflow.com/a/37843854 .

Fixed in 454fc272e0

thejayps commented:
proc.review.phase.request @rptb1 has requested this Pull Request be reviewed keybase://chat/jpholburn,rptb1,unaa008/1730, keybase://chat/jpholburn,rptb1,unaa008/1732

thejayps commented:
.express 
Changes are low risk. Main risk is lack of documentation visibility/clarity which is very unlikely to be mistaken for valid documentation.
This could be an express review, however I'm not able to quickly assess whether I'm a qualified express reviewer ([.role.leader.experienced](https://github.com/Ravenbrook/mps/blob/branch/2023-01-19/review-procedure/procedure/review.rst#role-leader-experienced)). It's also been a while since we did any review so I'll assume a full review is required.

thejayps commented:
Executing [proc.review.entry](https://github.com/Ravenbrook/mps/blob/586336a7021c2276a3f798ad10e6e4ab5f438226/procedure/review.rst#51-review-entry)

Start time 0005
[entry.impl](https://github.com/Ravenbrook/mps/blob/586336a7021c2276a3f798ad10e6e4ab5f438226/procedure/entry.impl.rst)
and [entry.universal](https://github.com/Ravenbrook/mps/blob/586336a7021c2276a3f798ad10e6e4ab5f438226/procedure/entry.universal.rst) apply

entry.impl.author: @rptb1 is the author, having made all the commits.
entry.impl.auto-check: all pass

.source-approved: The [source documents](https://github.com/Ravenbrook/mps/blob/586336a7021c2276a3f798ad10e6e4ab5f438226/procedure/review.rst#6-documents) will have [exited review](https://github.com/Ravenbrook/mps/blob/586336a7021c2276a3f798ad10e6e4ab5f438226/procedure/review.rst#58-review-exit). Failing this, they should be mini-reviewed and the fact noted. Failing this, they should be marked as "UNREVIEWED".

Assume no documents relating to this change have been reviewed. 
.rules-available
STOP - not clear what rules are relevant from https://github.com/Ravenbrook/mps/blob/586336a7021c2276a3f798ad10e6e4ab5f438226/procedure/entry.universal.rst

Which rules apply?

.entry paused

rptb1 commented:
> Which rules apply?

I suggest [rule.generic](https://github.com/Ravenbrook/mps/blob/586336a7021c2276a3f798ad10e6e4ab5f438226/procedure/rule.generic.rst) and [rule.style](https://github.com/Ravenbrook/mps/blob/586336a7021c2276a3f798ad10e6e4ab5f438226/procedure/rule.style.rst) . Possibly https://monzo.com/tone-of-voice/ . We have not developed any more detailed rules for the manual specifically.  Perhaps that something we can consider at [proc.review.ko.improve](https://github.com/Ravenbrook/mps/blob/586336a7021c2276a3f798ad10e6e4ab5f438226/procedure/review.rst#53-review-kickoff).

thejayps commented:
Resuming .entry at 1531:

rule.generic and rule.style apply. See https://github.com/Ravenbrook/mps/pull/166#issuecomment-1764055679 for more suggestions

Entry finished at 1536

thejayps commented:
Beginning [proc.review.planning](https://github.com/Ravenbrook/mps/blob/586336a7021c2276a3f798ad10e6e4ab5f438226/procedure/review.rst#52-review-planning)

Start time 1626

"Results are visible here https://memory-pool-system--166.org.readthedocs.build/en/166/ " (from keybase://chat/rptb1,jpholburn,unaa008/1736 )

Code is mostly rst, html, small amount of Python/make
70 loc (~= 7-10 mins at 7-10 loc/min)
1 Checker for consistency and correctness of code
1 Checker on correctness of OUTPUT (ie clearly displaying correct MPS information) uses search to make sure relevant pages are displayed
1 Checker checks issues are met and checks pages at random, via navigation, for clarity, including with previous manual for (e.g) rendering/formatting defects. Also checks "peripheral pages" eg indexes, links elsewhere

Checkers checking output should check 1 "page" in every 1-2 mins and tr

Total checking time 15 mins 

proc.review.plan.schedule: @thejayps, @rptb1 and @UNAA008 will review at a time TBC on 2023/10/19

Rules for style and generic apply. Checkers may also try @rptb1 's suggestion : https://github.com/Ravenbrook/mps/pull/166#issuecomment-1764055679 but the suggestion is not (yet) a reviewed source.

No homework.

Invite todo

Finished at 1648

thejayps commented:
Required for checkers:

"Results are visible here https://memory-pool-system--166.org.readthedocs.build/en/166/ " (from keybase://chat/rptb1,jpholburn,unaa008/1736 )

Old manual: visible to anyone following the link from   Ravenbrook home page: https://www.ravenbrook.com/project/mps/master/manual/html/

Old manual: visible to anyone following the link from MPS github page: 
https://memory-pool-system.readthedocs.io/en/latest/

Rules: 
https://github.com/Ravenbrook/mps/blob/586336a7021c2276a3f798ad10e6e4ab5f438226/procedure/rule.generic.rst
https://github.com/Ravenbrook/mps/blob/586336a7021c2276a3f798ad10e6e4ab5f438226/procedure/rule.style.rst

Other rules of type "code" found here would appear to apply, but it's possible these aren't appropriate for manual generation code: 
https://github.com/Ravenbrook/mps/tree/586336a7021c2276a3f798ad10e6e4ab5f438226/procedure

Other unnaproved, unreviewed rules suggested by @rptb1 https://monzo.com/tone-of-voice/


Checking procedure:
https://github.com/Ravenbrook/mps/blob/586336a7021c2276a3f798ad10e6e4ab5f438226/procedure/review.rst#54-review-checking

thejayps commented:
[executing proc.review.ko](https://github.com/Ravenbrook/mps/blob/586336a7021c2276a3f798ad10e6e4ab5f438226/procedure/review.rst#53-review-kickoff)

Started ~1415

After intro, JPH raised maintainability as an issue. @rptb1 explained risks of total collapse are slim. @UNAA008 highlights the requirement for others to be able to easily update documentation

metrics/special purposes:
Defects in review
Checking how quick/well we do a review after a review hiatus

JPH to prompt during follow up "how is this review stuff going"

Roles: 

(@UNAA008 highlights that output should be checked on a phone)

@UNAA008  Checker for consistency and correctness of code
@rptb1  Checker on correctness of OUTPUT (ie clearly displaying correct MPS information) uses search to make sure relevant pages are displayed
@thejayps  Checker checks issues are met and checks pages at random, via navigation, for clarity, including with previous manual for (e.g) rendering/formatting defects. Also checks "peripheral pages" eg indexes, links elsewhere and check output on phone

kickoff finished at 1538
Logging will start at 1600

rptb1 reviewed (commented):
Executing [proc.review.check](https://github.com/Ravenbrook/mps/blob/586336a7021c2276a3f798ad10e6e4ab5f438226/procedure/review.rst#542-checking).

1. Start time 15:40.
2. Started by assessing how well change solves linked issues.
3. Loaded manual onto phone (Fennec (Firefox) on LineageOS) and checked navigation, visibility of tables, readability of code samples, readability of designs, build instructions, function of search. #185 .
4. m. https://memory-pool-system--166.org.readthedocs.build/en/166/glossary/index.html It's hard to visually separate the terms in the navigation of the glossary.
5. Tested searching for various terms, some of which have come up in recent issues. Works well. #157 .
6. Ran `make -C manual clean; make -C manual html` in the branch on Ubuntu 22. Noted that Sphinx 7.2.6 was used, solving #121 .
7. M. The manual published at <https://www.ravenbrook.com/project/mps/master/manual/html/> is out of date and this will get worse.  Needs to be replaced by a redirect?
8. Switched to checking code changes.  2 minor defects found.
9. Totals: 1M, 3m.
10. Checking finished 15:59.

UNAA008 commented:
m1
requirements.pip header comment names the file as requirements.txt

thejayps reviewed (commented):
m1. Search doesn't have case sensitive option on phone (eg, search for "Pin" found "MutatorContextCanSte*pIn*struction")

m2. Search doesn't appear on phone unless you select the "burger" (but maybe this is appropriate and expected by phone users)

c1. The pull request could do more to explain exactly how it fixes #121 

c2. The output contains ads "Ads by EthicalAds" 
![image](https://github.com/Ravenbrook/mps/assets/73122241/3963d6b6-4d41-4a9e-89a8-be4178fadb38)

m3. "Abstract" in bibliography appears with a "!" like a warning

UNAA008 commented:
Q1
Do we need a different mechanism to implement rule.generic.ident ?

thejayps commented:
executing [proc.review.log](https://github.com/Ravenbrook/mps/blob/586336a7021c2276a3f798ad10e6e4ab5f438226/procedure/review.rst#55-review-logging)

Start time 1610

@rptb1 new.M2: What was found in item 7 also applies to everything under master
@UNAA008 new.m In response to m2 by @thejayps , convention is also to have a magnifier for search, we don't have one

@rptb1 raises new.minor.imp Git wont automatically stamp our files with a source control revision ID so we fail to meet rule.genric.ident. There's other sorts of ident (eg product document) that is in a way more important

@rptb1 new.m in "contributing to the mps" the text is stale. We could do better eg reference to the branch.

logging finised at 1649

rptb1 commented:
> 7\. M. The manual published at https://www.ravenbrook.com/project/mps/master/manual/html/ is out of date and this will get worse.  Needs to be replaced by a redirect?

Just noting that this is relevant to issue #98 in that the whole tree below https://www.ravenbrook.com/project/mps/ is a manifestation of the Perforce project, and much of it should be redirected.

rptb1 commented:
> c2. The output contains ads "Ads by EthicalAds"

Subscribed Ravenbrook to "Gold Membership" of Read the Docs.  The ad status of projects can be managed at https://readthedocs.org/accounts/gold/projects/ .  I have disabled ads on the MPS.

rptb1 commented:
> m1. Search doesn't have case sensitive option on phone (eg, search for "Pin" found "MutatorContextCanSte_pIn_struction")

Raise: https://github.com/readthedocs/sphinx_rtd_theme/issues/1534

rptb1 commented:
> m2. Search doesn't appear on phone unless you select the "burger" (but maybe this is appropriate and expected by phone users)

Reject: It is appropriate, I think. You can also rotate the phone :)

rptb1 commented on manual/source/conf.py:
```diff
@@ -37,7 +37,13 @@
-    html_theme = 'mps'
+    # We use the Read the Docs theme for Sphinx for smooth publishing
+    # at readthedocs.io and low maintenance overheads.  See
+    # <https://sphinx-rtd-theme.readthedocs.io/>.
```
Fix: Added references in 421cad4e6

rptb1 commented on manual/source/_templates/layout.html:
```diff
@@ -0,0 +1,32 @@
+
+<ul>
+
+  <li class="toctree-11">
```
It's not even correct.  It should be toctree-l1.
Fix: Corrected and explained in 6fd0a9110 .

rptb1 commented:
> c1. The pull request could do more to explain exactly how it fixes #121

Reply: Discussed during review, but documenting here: We don't know exactly what is wrong with our homegrown theme in #121 and it would be wasteful to debug it, since we've chosen to switch to the Read the Docs theme.  In short, #121 is fixed by using a community-maintained up-to-date theme.

rptb1 commented:
> Q1
> Do we need a different mechanism to implement rule.generic.ident ?

Raise: https://github.com/Ravenbrook/mps/issues/98#issuecomment-1772750581

rptb1 commented:
> @rptb1 new.m in "contributing to the mps" the text is stale. We could do better eg reference to the branch.

Raise: This should be fixed as part of #123 .

rptb1 commented:
Executing [proc.review.edit](https://github.com/Ravenbrook/mps/blob/586336a7021c2276a3f798ad10e6e4ab5f438226/procedure/review.rst#57-review-edit).

1. Start time 12:15.
2. Written responses submitted to #166 .
3. End time 14:46.

thejayps reviewed (approved):
"Approving" after checking edits as requested by @rptb1 

ready for exit

thejayps commented:
performing [proc.review.exit](https://github.com/Ravenbrook/mps/blob/586336a7021c2276a3f798ad10e6e4ab5f438226/procedure/review.rst#58-review-exit)

Start time 1830

Transgressions:

T1 exit.universal.edit: "m3. "Abstract" in bibliography appears with a "!" like a warning" - there is no action recorded 

T2 exit.universal.imp: "new.minor.imp Git wont automatically stamp our files with a source control revision ID so we fail to meet rule.genric.ident. There's other sorts of ident (eg product document) that is in a way more important" - there is no action recorded

T3: Fixed: exit.universal.record estimation of remaining defects is missing. Estimating 1M remains, as 2M were found

(not a transgression) exit.universal.rates - rates were in danger of being too fast as the estimation of LOC was based only on the diffs. However, relevant checkers took more time and then reported having enough time to do the checking properly

Exit paused at 1847 while transgressions T1-2 discussed

rptb1 commented:
> T2 exit.universal.imp: "new.minor.imp Git wont automatically stamp our files with a source control revision ID so we fail to meet rule.genric.ident. There's other sorts of ident (eg product document) that is in a way more important" - there is no action recorded

Included in https://github.com/Ravenbrook/mps/pull/166#issuecomment-1772751148 but I failed to link it.

> T1 exit.universal.edit: "m3. "Abstract" in bibliography appears with a "!" like a warning" - there is no action recorded

Genuinely overlooked that one.  I'll see if there's some better markup.

rptb1 commented:
> > T1 exit.universal.edit: "m3. "Abstract" in bibliography appears with a "!" like a warning" - there is no action recorded
> 
> Genuinely overlooked that one. I'll see if there's some better markup.

Reject: Cosmetic and too expensive to fix.  I've investigated how this comes about and overriding it would be tricky, fragile, and costly.  The markup is semantically probably correct (a custom admonition).  We may come up with something later in passing.

rptb1 commented:
> 4\. m. https://memory-pool-system--166.org.readthedocs.build/en/166/glossary/index.html It's hard to visually separate the terms in the navigation of the glossary.

I'm concerned that I missed this, and @thejayps didn't find it at exit.  It suggests that [proc.review.log](https://github.com/Ravenbrook/mps/blob/586336a7021c2276a3f798ad10e6e4ab5f438226/procedure/review.rst#55-review-logging) isn't producing a findable log.

rptb1 commented:
> 4\. m. https://memory-pool-system--166.org.readthedocs.build/en/166/glossary/index.html It's hard to visually separate the terms in the navigation of the glossary.

Fixed: Introduced punctuation and spacing in https://github.com/Ravenbrook/mps/pull/166/commits/ebeee66005d5c46621e0b2707acc91bfecc7e57c

rptb1 commented:
I believe that concludes [proc.review.exit.fix](https://github.com/Ravenbrook/mps/blob/586336a7021c2276a3f798ad10e6e4ab5f438226/procedure/review.rst#58-review-exit) so merge is pending a pass from @thejayps .

thejayps commented:
exit.pass

rptb1 commented:
Executing [proc.merge.pull-request](https://github.com/Ravenbrook/mps/blob/554513b16461ec0bbf82856dfe6b0528518c3658/procedure/pull-request-merge.rst).

1. Start time 10:07.
2. Checklist: The test cases for the issues are not automated: #185 is subjective, #157 is UI (though we could conceivably automate), #121 is a build failure.
3. Skipping steps to do with Git Fusion.
4. End time 10:14.

rptb1 commented:
> 3\. Skipping steps to do with Git Fusion.

We should get on with https://github.com/Ravenbrook/mps/pull/228 to fix this.
